### ISSUE TITLE
RecursionError in gcd for unevaluated Mul func with integer args

### ISSUE THREAD

**Issue by bsdz** [sympy/sympy#20597]
Spotted this earlier. I have workaround by just pre-evaluating before running gcd.

```python
>>> import sympy as sp
>>> sp.gcd(sp.Mul(2, 3, evaluate=False), 2)
---------------------------------------------------------------------------
PolificationFailed                        Traceback (most recent call last)
.../.venv/lib/python3.7/site-packages/sympy/polys/polytools.py in gcd(f, g, *gens, **args)
   5277     try:
-> 5278         (F, G), opt = parallel_poly_from_expr((f, g), *gens, **args)
   5279 

.../.venv/lib/python3.7/site-packages/sympy/polys/polytools.py in parallel_poly_from_expr(exprs, *gens, **args)
   4311     opt = options.build_options(gens, args)
-> 4312     return _parallel_poly_from_expr(exprs, opt)
   4313 

.../.venv/lib/python3.7/site-packages/sympy/polys/polytools.py in _parallel_poly_from_expr(exprs, opt)
   4366     if not opt.gens:
-> 4367         raise PolificationFailed(opt, origs, exprs, True)
   4368 

PolificationFailed: can't construct polynomials from 2*3, 2

During handling of the above exception, another exception occurred:

<stack repeats then eventually...>

RecursionError: maximum recursion depth exceeded
```

Using another operator works fine:

```python
>>> sp.gcd(sp.Pow(2, 3, evaluate=False), 2)
2
```

**Comment by oscarbenjamin** [comment]
Are you using sympy 1.6.2? I get a different result in 1.7.1:
```python
In [1]: gcd(Mul(2, 3, evaluate=False), 2)
Out[1]: 1
```
Obviously that should be 2:
```python
In [2]: gcd(6, 2)
Out[2]: 2

In [3]: gcd(Mul(2, 3, evaluate=False).doit(), 2)
Out[3]: 2
```
The root of the problem is that `gcd` converts expressions to Poly which tries to convert expressions into a standard form using `expand`. When the fails to produce the expected results gcd ends up computing over the EX domain:
```python
In [5]: expand(Mul(2, 3, evaluate=False))
Out[5]: 2⋅3

In [6]: expand(Pow(2, 3, evaluate=False))
Out[6]: 8

In [7]: expand(Add(2, 3, evaluate=False))
Out[7]: 2 + 3
```
That suggests we would see the same problem with `Add` which we do:
```python
In [9]: gcd(Add(3, 3, evaluate=False), 2)
Out[9]: 1
```
This is the function that should handle this for `Mul`:
https://github.com/sympy/sympy/blob/0bde5c951b52341e745d5410dba1a9a75d7acebf/sympy/core/mul.py#L862-L889
That could be fixed to make `expand` work for unevaluated `Mul`. It seems that it works if there is anything else to expand:
```python
In [2]: Mul(2, 3, (1 + x), evaluate=False).expand()
Out[2]: 6⋅x + 6
```
The problem is at line 887 where `expand` concludes that there is nothing to do if there was no `Add` in its args.

I guess something similar happens in `Add._eval_expand_mul`.

**Comment by bsdz** [comment]
> Are you using sympy 1.6.2? I get a different result in 1.7.1:

Yes that's right. I was using 1.6.2 and now updated to 1.7.1 and see same as you (ie 1).

> The problem is at line 887 where `expand` concludes that there is nothing to do if there was no `Add` in its args.
> 
> I guess something similar happens in `Add._eval_expand_mul`.

Sounds like a good explanation. My workaround was to doit() on each of the parameters in gcd(). That works for me since, as right now, all my args are integers. I'm don't feel using doit() would be a good general solution. Your suggestion is definitely the way to go.

**Comment by AaronStiff** [comment]
Hi @oscarbenjamin . I'm new to this project and was wondering if you could help me understand the code structure (not sure if this is the right place to ask). From what I can see `sympy.gcd()` is a function of `Poly`. You said:

> `gcd` converts expressions to Poly which tries to convert expressions into a standard form using `expand`. When the fails to produce the expected results gcd ends up computing over the EX domain:

However, when I look at the definition of `gcd`, I don't see where it is doing this. I do see `_unify` though (which for some reason is only accepting one argument when it requires 2). Am I missing something further down? 

Also, is this issue simple enough for a first project? Thanks!

**Comment by oscarbenjamin** [comment]
We can find out the module containing the `gcd` function from its `repr`:
```
In [1]: gcd
Out[1]: <function sympy.polys.polytools.gcd(f, g=None, *gens, **args)>
```
So we see that it's in `sympy/polys/polytools.py`. We can go there and search for `def gcd` which takes us to
https://github.com/sympy/sympy/blob/5288eb336d12e4f3c3070bce67d75f38ab4d8376/sympy/polys/polytools.py#L5261-L5286

At line 5286 we see the call to `parallel_poly_from_expr` which tries to convert both expressions to `Poly`.

I guess that the code that parses the expression to convert identify the domain could also be improved to handle this without using `expand` but that's a more difficult fix. I have been meaning to rewrite the internals of `construct_domain` so that they recursively deconstruct and reconstruct the expressions. I already did that for `construct_algebraic` in #20274 but we could rewrite the whole of `construct_domain` to work that way:
https://github.com/sympy/sympy/blob/5288eb336d12e4f3c3070bce67d75f38ab4d8376/sympy/polys/constructor.py#L81-L132

It's probably not worth following that line through the polys code if we can fix this in the expand code though. If expand worked then `gcd` would work in these cases so the place to look is `Mul._eval_expand_mul` (as I showed above) I think. The logic there should be adjusted somehow.

> Also, is this issue simple enough for a first project? Thanks!

There are certainly easier issues to work on but fixing `expand` here might be reasonable if you can follow what the code there is doing and that will fix the basic issue here.

Fixing this more generally in the polys code as I suggested above is probably a lot harder just because the internals there are less user friendly - the fix there would be extremely useful though because calling expand when constructing polynomials is a big slowdown in many cases.

**Comment by AaronStiff** [comment]
Great, thanks! Just out of curiousity, is that a special terminal you used `repr` with, and if so, how would I do that from a command line?

Being new, I'll probably just see if I can fix `expand`. I assume I should let you know if I don't make any headway into it?

**Comment by oscarbenjamin** [comment]
I'm using `isympy`. If you have sympy installed you should be able to run `isympy` from the command line:
```console
$ isympy
IPython console for SymPy 1.8.dev (Python 3.8.5-64-bit) (ground types: python)

These commands were executed:
>>> from __future__ import division
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
>>> init_printing()

Documentation can be found at https://docs.sympy.org/dev


In [1]: Integral(x**2, (x, 0, 1))
Out[1]: 
1      
⌠      
⎮  2   
⎮ x  dx
⌡      
0   
```
The result comes out a lot prettier in a Linux/OSX terminal than it does in the standard Windows console.

> Being new, I'll probably just see if I can fix `expand`. I assume I should let you know if I don't make any headway into it?

Yes, feel free to ask for help or otherwise to report here on anything that does or doesn't work as that can also be helpful if someone else takes this up later.

**Comment by AaronStiff** [comment]
It feels kind of like a hack, but adding this in just before line 887 in `_eval_expand_mul` seems to work:

```python
if plain and not sums:
    rewrite = True
    for p in plain:
        if not p.is_integer:
            rewrite = False
```

It's essentially checking to see if the expression is a `Mul` with only integers (the case causing the bug) and forcing a rewrite.

Again, because I'm not very familiar with the project as a whole, I'd welcome feedback as to whether this breaks something obvious.

**Comment by oscarbenjamin** [comment]
Note that `is_integer` is different from `is_Integer`. The first is a semantic check as part of the assumptions system whereas the second as a structural check for a literal `Integer` in an expression:
```
In [3]: x = Symbol('x', integer=True)

In [4]: x.is_integer
Out[4]: True

In [5]: x.is_Integer
Out[5]: False

In [6]: S(2).is_integer
Out[6]: True

In [7]: S(2).is_Integer
Out[7]: True
```

The easiest way to check if this breaks anything is to run the test suite. Alternatively you can open a pull request and the test suite will run in CI.

I wonder if we could make use of `n` and `d` somehow. I think that `fraction` already handles combining the unevaluated Mul or am I mistaken?

**Comment by AaronStiff** [comment]
I see. So would it be preferable to use `is_Integer` instead?

> I think that fraction already handles combining the unevaluated Mul or am I mistaken?

You are correct. I noticed it as I was testing. It takes the expression `sp.Mul(2, 3, evaluate = False)` for instance and returns `n = 6` and `d = 1`. Were you thinking something like this instead (to be added after line 868):

```python
if n.is_Integer:
    expr = n
```

**Comment by oscarbenjamin** [comment]
Or perhaps just `expr = n/d`.

**Comment by AaronStiff** [comment]
Yeah, that's perfect. I'll create a pull request, but I'm not really sure I can take credit for it. 😃 

Thanks for the clarity you brought to the issue (e.g. tracing the issue to `_eval_expand_mul`). If nothing else, I've gained a fair bit of knowledge. 👍

**Comment by oscarbenjamin** [comment]
See also https://github.com/sympy/sympy/issues/20715#issuecomment-753604519 which I think is the same issue.

**Comment by povinsahu1909** [comment]
In `gcd(Mul(2, 3, evaluate=False), 2)` and `gcd(Add(3, 3, evaluate=False), 2)` the domain from line 5297 is `EX`
https://github.com/sympy/sympy/blob/8ee5e72726616e6ee8ef9ebfced5e1904e2a8167/sympy/polys/polytools.py#L5285-L5303
which then calls the following:
https://github.com/sympy/sympy/blob/8ee5e72726616e6ee8ef9ebfced5e1904e2a8167/sympy/polys/domains/expressiondomain.py#L226-L230
which is returns 1.
But if we replace line 227 to `return a.gcd(b)` (similar to what we have for lcm on line 230) it calls the following function
https://github.com/sympy/sympy/blob/8ee5e72726616e6ee8ef9ebfced5e1904e2a8167/sympy/polys/domains/expressiondomain.py#L125-L127
which gives `RecursionError` as it again calls gcd.

But for all other cases (as far as I know) where domain is `QQ` or `ZZ`, calls the following function which uses GMPY gcd function to calculate the gcd 
https://github.com/sympy/sympy/blob/8ee5e72726616e6ee8ef9ebfced5e1904e2a8167/sympy/polys/domains/gmpyintegerring.py#L81-L83.
So can we change the gcd function in `sympy/sympy/polys/domains/expressiondomain.py` similar to other domains which were dependent on GMPY to avoid this error.

**Comment by povinsahu1909** [comment]
There is similar RecursionError while executing `lcm(Mul(2,3, evaluate = False) , 6) ` due to following code
https://github.com/sympy/sympy/blob/8ee5e72726616e6ee8ef9ebfced5e1904e2a8167/sympy/polys/domains/expressiondomain.py#L229-L230

**Comment by povinsahu1909** [comment]
What I think is that the main problem here is somehow related to the domain of the numbers passed.
And should I open another issue regarding similar error in lcm function as commented above?

**Comment by oscarbenjamin** [comment]
I don't think we need a separate issue. This one can remain open until each issue is fixed.

The problems for both `lcm` and `gcd` stem from the problem with `expand` when used with unevaluated `Add/Mul`.
